### PR TITLE
feat: Add advanced email validation methods to Email rule

### PR DESCRIPTION
**Pull request by MahdiBagheri71**
## Description

This PR enhances the Laravel Email validation rule by adding advanced email validation methods while maintaining full backward compatibility. These methods address common real-world email validation challenges that developers face when building secure web applications.

## Problem Statement

Current Laravel email validation provides basic functionality, but many applications require advanced email validation to:
- Prevent fake/temporary email registrations
- Block email aliases that bypass business rules
- Ensure email deliverability through MX validation
- Protect against sophisticated email-based attacks
- Enforce company email policies

## Solution

Added 9 new optional methods to the `Illuminate\Validation\Rules\Email` class:

### Core Validation Methods
- **`validateMxRecord()`** - Validates DNS/MX records to ensure email deliverability
- **`preventSpoofing()`** - Prevents unicode spoofing attacks in email addresses
- **`rfcCompliant(bool $strict = false)`** - Enhanced RFC compliance validation
- **`withNativeValidation(bool $allowUnicode = false)`** - PHP native validation integration

### Advanced Protection Methods
- **`noDots()`** - Blocks Gmail dot aliases ([user.name@gmail.com](mailto:user.name@gmail.com) → [username@gmail.com](mailto:username@gmail.com))
- **`noAliases()`** - Blocks common aliasing techniques ([user+tag@domain.com](mailto:user+tag@domain.com))
- **`trustedDomains(?array $domains = null)`** - Restricts to whitelisted domains only
- **`noDisposable()`** - Blocks temporary email services (tempmail.org, guerrillamail.com, etc.)
- **`noForwarding()`** - Blocks email forwarding services (simplelogin.io, anonaddy.com, etc.)
- **`noSuspiciousPatterns()`** - Blocks suspicious patterns (test123, temp456, etc.)

### Convenience Methods
- **`strictAdvanced()`** - Enables all advanced protections with one method
- **`strict()`** - Shorthand for strict RFC compliance

## Benefits to End Users

1. **Enhanced Security**: Prevents fake registrations and email-based attacks
2. **Better Data Quality**: Ensures valid, deliverable email addresses
3. **Flexible Configuration**: Choose exactly which validations your app needs
4. **Zero Breaking Changes**: Completely opt-in, existing code continues working
5. **Fluent API**: Chain methods for readable validation rules
6. **Business Logic Enforcement**: Block aliases and temporary emails as needed

## Real-World Use Cases

```php
// User registration with deliverability check
Email::default()->validateMxRecord()->noDisposable()

// Corporate application requiring company emails
Email::default()->trustedDomains(['company.com', 'partner.org'])

// High-security application
Email::default()->strictAdvanced()->validateMxRecord()

// Newsletter preventing aliases
Email::default()->noAliases()->noDisposable()
```

## Why It Doesn't Break Existing Features

- **100% Backward Compatible**: All existing `Email` rule usage continues unchanged
- **Opt-in Only**: New methods must be explicitly called
- **Existing Tests Pass**: All current validation behavior preserved
- **Same Interface**: Implements same contracts (`Rule`, `DataAwareRule`, `ValidatorAwareRule`)
- **Message System**: Uses existing Laravel validation message infrastructure

## How It Makes Building Web Applications Easier

1. **Reduces Custom Code**: No need to write custom validation classes for common needs
2. **Framework Consistency**: Uses Laravel's familiar fluent validation syntax
3. **Comprehensive Solution**: Covers most email validation scenarios out-of-the-box
4. **Easy Integration**: Drop-in replacement with additional capabilities
5. **Maintainable**: Framework-level maintenance vs. scattered custom solutions

## Implementation Details

- **Performance Optimized**: Domain checks use fast array lookups
- **Extensible**: Easy to add new disposable/forwarding domains
- **Well-Documented**: Comprehensive PHPDoc comments
- **Error Handling**: Descriptive validation messages for each failure type
- **Memory Efficient**: Lazy loading of validation rules

## Testing

- **Comprehensive Coverage**: Tests for all new methods and edge cases
- **Existing Tests**: All current tests continue to pass
- **Integration Tests**: Works with Laravel's validation system
- **Error Message Tests**: Validates proper error messaging

## Examples

```php
// Basic upgrade path
// Before
'email' => 'required|email'

// After
'email' => ['required', Email::default()]

// Advanced usage
'email' => [
    'required',
    Email::default()
        ->validateMxRecord()
        ->noDisposable()
        ->preventSpoofing()
        ->rules('unique:users,email')
]
```

## Supported Use Cases

- **SaaS Applications**: Prevent fake signups and improve data quality
- **Enterprise Apps**: Enforce domain restrictions and security policies
- **E-commerce**: Ensure customer emails are deliverable for order notifications
- **Marketing Platforms**: Block temporary emails that hurt campaign metrics
- **Community Platforms**: Prevent ban evasion through email aliases

This enhancement makes Laravel's email validation more powerful while maintaining the framework's principle of elegant, expressive code.

### CODE REVIEW

File: src/Illuminate/Validation/Rules/Email.php [hunk 1]
```diff
@@ -162,6 +169,93 @@ public function withNativeValidation(bool $allowUnicode = false)
...
+     *
+     * @param  array|null  $domains
+     * @return $this
+     */
+    public function trustedDomains(?array $domains = null)
+    {
+        $this->trustedDomainsOnly = true;
+        if ($domains !== null) {
+            $this->customTrustedDomains = $domains;
+        }
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:222]: You could shorten this if wanted
```suggestion
        $this->trustedDomainsOnly = $domains !== null
            ? $domains
            : true;
```
or even
```suggestion
        $this->trustedDomainsOnly = $domains === null || $domains;
```

File: src/Illuminate/Validation/Rules/Email.php [hunk 2]
```diff
@@ -190,16 +284,107 @@ public function passes($attribute, $value)
...
+        // Run standard Laravel email validation (except DNS since we handle it above)
+        $rules = $this->buildValidationRules();
+
+        // Remove dns from rules if we're handling it manually
+        if ($this->validateMxRecord) {
+            $rules = array_map(function ($rule) {
+                if (str_starts_with($rule, 'email:')) {
+                    $parts = explode(':', $rule);
+                    if (isset($parts[1])) {
+                        $flags = explode(',', $parts[1]);
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:320]: You could go without array indexes
```suggestion
                    [, $flagList] = explode(':', $rule);
                    if (isset($flagList)) {
                        $flags = explode(',', $flagList);
```
or simply
```suggestion
                    $flagList = strstr($rule, ':', true);
                    if (isset(flagList)) {
                        $flags = explode(',', flagList);
```

File: src/Illuminate/Validation/Rules/Email.php [hunk 3]
```diff
@@ -190,16 +284,107 @@ public function passes($attribute, $value)
...
+                        $flags = array_filter($flags, fn($f) => $f !== 'dns');
+                        return $flags ? 'email:' . implode(',', $flags) : 'email';
+                    }
+                }
+                return $rule;
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:325]: You might use early returns here:
```suggestion
                if (! str_starts_with($rule, 'email:')) {
                    return $rule;
                }

                $parts = explode(':', $rule);
                if (isset($parts[1])) {
                    return $rule;
                }

                $flags = explode(',', $parts[1]);
                $flags = array_filter($flags, fn($f) => $f !== 'dns');
                return $flags ? 'email:' . implode(',', $flags) : 'email';
```

File: src/Illuminate/Validation/Rules/Email.php [hunk 4]
```diff
@@ -190,16 +284,107 @@ public function passes($attribute, $value)
...
+
+        // Run standard Laravel email validation (except DNS since we handle it above)
+        $rules = $this->buildValidationRules();
+
+        // Remove dns from rules if we're handling it manually
+        if ($this->validateMxRecord) {
+            $rules = array_map(function ($rule) {
+                if (str_starts_with($rule, 'email:')) {
+                    $parts = explode(':', $rule);
+                    if (isset($parts[1])) {
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:319]: Will this ever not be true?

File: src/Illuminate/Validation/Rules/Email.php [hunk 5]
```diff
@@ -190,16 +284,107 @@ public function passes($attribute, $value)
...
+
+        if ($this->blockForwarding && $this->isForwardingService($domain)) {
+            $this->messages[] = 'Email forwarding services are not allowed.';
+            return false;
+        }
+
+        if ($this->blockSuspiciousPatterns && $this->hasSuspiciousPatterns($localPart)) {
+            $this->messages[] = 'This email format appears to be temporary or invalid.';
+            return false;
+        }
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:384]: Can't you use the messages from the validation translation file here?

File: src/Illuminate/Validation/Rules/Email.php [hunk 6]
```diff
@@ -283,4 +468,175 @@ public function setData($data)
...
+            'yahoo.com.au',
+            'yahoo.de',
+            'yahoo.fr',
+            'yahoo.es',
+            'yahoo.it',
+            'ymail.com',
+            'rocketmail.com',
+            'aol.com',
+            'aim.com',
+        ];
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:505]: What makes them especially trusted? Is this a custom list or does it come from somewhere official?

File: src/Illuminate/Validation/Rules/Email.php [hunk 7]
```diff
@@ -283,4 +468,175 @@ public function setData($data)
...
+     * @param  string  $localPart
+     * @return bool
+     */
+    protected function hasAliases(string $localPart): bool
+    {
+        return str_contains($localPart, '+') ||
+            preg_match('/\.{2,}/', $localPart) ||
+            str_starts_with($localPart, '.') ||
+            str_ends_with($localPart, '.');
+    }
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:592]: I'm not sure if we want the gmail dots to be provider-specific while we don't apply the aliases solely to providers that actually implement that part of the spec

File: src/Illuminate/Validation/Rules/Email.php [hunk 8]
```diff
@@ -283,4 +468,175 @@ public function setData($data)
...
+     * @return bool
+     */
+    protected function hasSuspiciousPatterns(string $localPart): bool
+    {
+        $suspiciousPatterns = [
+            '/^[a-z]\d{8,}$/', // Random letters + numbers
+            '/^test\d*$/', // Test emails
+            '/^temp\d*$/', // Temporary emails
+            '/^noreply/', // No-reply emails
+            '/^admin\d*$/', // Admin emails
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:629]: Why is `admin@example.com` suspicious?

File: src/Illuminate/Validation/Rules/Email.php [hunk 9]
```diff
@@ -283,4 +468,175 @@ public function setData($data)
...
+    /**
+     * Check for suspicious patterns.
+     *
+     * @param  string  $localPart
+     * @return bool
+     */
+    protected function hasSuspiciousPatterns(string $localPart): bool
+    {
+        $suspiciousPatterns = [
+            '/^[a-z]\d{8,}$/', // Random letters + numbers
```
Comment (shaedrich) [inline src/Illuminate/Validation/Rules/Email.php:625]: Why specifically 8?

File: tests/Validation/ValidationEmailRuleTest.php [hunk 10]
```diff
@@ -347,6 +347,379 @@ public function testRfcCompliantNonStrict()
...
+        $this->passes(
+            (new Email())->noAliases(),
+            'user@gmail.com'
+        );
+
+        $this->passes(
+            Rule::email()->noAliases(),
+            'validuser@gmail.com'
+        );
+    }
```
Comment (shaedrich) [inline tests/Validation/ValidationEmailRuleTest.php:427]: You could use a [data provider](https://docs.phpunit.de/en/10.5/writing-tests-for-phpunit.html#writing-tests-for-phpunit-data-providers) here
```suggestion
    #[DataProvider('provideDotAliasRules')]
    public function testNoAliases(Email $rule, string $emailAddress, ?array $errorMessages = null)
    {
    	if ($errorMessages === null) {
    		$this->passes($rule, $emailAddress);
    	} else {
	        $this->fails($rule, $emailAddress, $errorMessages);
    	}
    }
    
    public static function provideDotAliasRules()
    {
        return [
        	'email class and gmail address with plus' => [
            	(new Email())->noAliases(),
            	'user+test@gmail.com',
            	['Email aliases are not allowed.'],
        	},
        	'rule method and gmail address with plus' => [
	            (new Email())->noAliases(),
	            'user+test@gmail.com',
	            ['Email aliases are not allowed.']
        	],
			'email class and gmail address with dots within' => [
	            (new Email())->noAliases(),
	            'user..name@gmail.com',
	            ['Email aliases are not allowed.']
			],
			'email class and gmail address with leading dots' => [
	            (new Email())->noAliases(),
	            '.username@gmail.com',
	            ['Email aliases are not allowed.']
			],
			'email class and gmail address with trailing dots' => [
	            (new Email())->noAliases(),
            	'username.@gmail.com',
	            ['Email aliases are not allowed.']
			],
			'email class and gmail address without pluses and dots' => [
	            (new Email())->noAliases(),
            	'username.@gmail.com',
			],
			'rule method and gmail address without pluses and dots' => [
	            Rule::email()->noAliases(),
	            'validuser@gmail.com'
			],
        ];
    }
```
or
```suggestion
    #[DataProvider('providePassingDotAliasRules')]
    public function testNoAliases(Email $rule, string $emailAddress, array $errorMessages = null)
    {
    	$this->passes($rule, $emailAddress);
    }
    
    public static function providePassingDotAliasRules()
    {
        return [
			'email class and gmail address without pluses and dots' => [
	            (new Email())->noAliases(),
            	'username.@gmail.com',
			],
			'rule method and gmail address without pluses and dots' => [
	            Rule::email()->noAliases(),
	            'validuser@gmail.com'
			],
        ];
    }
    #[DataProvider('provideFailingDotAliasRules')]
    public function testNoAliases(Email $rule, string $emailAddress, array $errorMessages = null)
    {
	    $this->fails($rule, $emailAddress, $errorMessages);
    }
    
    public static function provideFailingDotAliasRules()
    {
        return [
        	'email class and gmail address with plus' => [
            	(new Email())->noAliases(),
            	'user+test@gmail.com',
            	['Email aliases are not allowed.'],
        	},
        	'rule method and gmail address with plus' => [
	            (new Email())->noAliases(),
	            'user+test@gmail.com',
	            ['Email aliases are not allowed.']
        	],
			'email class and gmail address with dots within' => [
	            (new Email())->noAliases(),
	            'user..name@gmail.com',
	            ['Email aliases are not allowed.']
			],
			'email class and gmail address with leading dots' => [
	            (new Email())->noAliases(),
	            '.username@gmail.com',
	            ['Email aliases are not allowed.']
			],
			'email class and gmail address with trailing dots' => [
	            (new Email())->noAliases(),
            	'username.@gmail.com',
	            ['Email aliases are not allowed.']
			],
        ];
    }
```

**Comment by taylorotwell** [comment]
Any further email validation should just be a package imo.
